.github/workflows/cd.yml

```diff
@@ -37,24 +37,54 @@ jobs:
             echo "SERVICE_NAME=${{ vars.AWS_ECS_API_SERVICE_STAGING }}" >> $GITHUB_ENV
           fi
 
-      - name: Set up QEMU
-        uses: docker/setup-qemu-action@v2
+      - name: Check for existing image with GIT_SHA
+        env:
+          GIT_SHA: ${{ github.sha }}
+          REPO: ${{ vars.AWS_ECR_API_REPO }}
+        run: |
+          image_manifest=$(aws ecr batch-get-image --repository-name $REPO --image-ids imageTag=$GIT_SHA --query 'images[0].imageManifest')
+          echo "IMAGE_MANIFEST=$image_manifest" >> $GITHUB_ENV
+
+      - if: env.IMAGE_MANIFEST == 'null'
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - if: env.IMAGE_MANIFEST == 'null'
+        name: Build, tag, and push docker image to Amazon ECR
+        uses: docker/build-push-action@v5
         with:
-          platforms: arm64
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          context: .
+          file: ./packages/daimo-api/Dockerfile
+          platforms: linux/arm64
+          push: true
+          tags: |
+            ${{ steps.login-ecr.outputs.registry }}/${{ vars.AWS_ECR_API_REPO }}:${{ github.sha }}
+            ${{ steps.login-ecr.outputs.registry }}/${{ vars.AWS_ECR_API_REPO }}:${{ env.IMAGE_TAG }}
 
-      - name: Build, tag, and push docker image to Amazon ECR
+      - if: env.IMAGE_MANIFEST != 'null'
+        name: Add tag to existing image.
         env:
-          GIT_SHA: ${{ github.sha }}
-          IMAGE_NAME: ${{ steps.login-ecr.outputs.registry }}/${{ vars.AWS_ECR_API_REPO }}
+          REPO: ${{ vars.AWS_ECR_API_REPO }}
         run: |
-          docker build --platform linux/arm64 -t $IMAGE_NAME:$GIT_SHA -t $IMAGE_NAME:$IMAGE_TAG -f packages/daimo-api/Dockerfile .
-          docker push $IMAGE_NAME:$GIT_SHA
-          docker push $IMAGE_NAME:$IMAGE_TAG
+          echo "Image exists with GIT_SHA tag, adding tag \"$IMAGE_TAG\" to it."
+          IMAGE_MANIFEST=$(echo $IMAGE_MANIFEST | jq -c -r)
+
+          set +e
+          output=$(aws ecr put-image --repository-name $REPO --image-tag $IMAGE_TAG --image-manifest "$IMAGE_MANIFEST" 2>&1)
+          status=$?
+          set -e
+
+          if [[ $status -ne 0 && ! "$output" =~ "ImageAlreadyExistsException" ]]; then
+            echo $output
+            exit $status
+          else
+            echo "Image already tagged with \"$IMAGE_TAG.\""
```

Comment:
this will print "Image already tagged..." even if the command was successful, right? when $status = 0

Reply:
good catch! meant to have that nested under the non-zero catch. will address